src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs

```diff
@@ -22,23 +22,32 @@ private static unsafe int WaitForMultipleObjectsIgnoringSyncContext(IntPtr* pHan
         {
             Debug.Assert(millisecondsTimeout >= -1);
 
-            //
-            // In the CLR, we use CoWaitForMultipleHandles to pump messages while waiting in an STA.  In that case, we cannot use WAIT_ALL.
-            // That's because the wait would only be satisfied if a message arrives while the handles are signalled.
-            //
-            if (waitAll)
+            // Normalize waitAll
+            if (numHandles == 1)
+                waitAll = false;
+
+            bool reentrantWait = Thread.ReentrantWaitsEnabled;
+
+            if (reentrantWait)
             {
-                if (numHandles == 1)
-                    waitAll = false;
-                else if (Thread.GetCurrentApartmentType() == Thread.ApartmentType.STA)
+                //
+                // In the CLR, we use CoWaitForMultipleHandles to pump messages while waiting in an STA.  In that case, we cannot use WAIT_ALL.
+                // That's because the wait would only be satisfied if a message arrives while the handles are signalled.
+                //
+                if (waitAll)
                     throw new NotSupportedException(SR.NotSupported_WaitAllSTAThread);
+
+                // CoWaitForMultipleHandles does not support more than 63 handles. It returns RPC_S_CALLPENDING for more than 63 handles
+                // that is impossible to differentiate from timeout.
+                if (numHandles > 63)
```

Comment:
This is fixing a test failure that results into fatal unhandled in System.Runtime.Tests run

Reply:
Wouldn't even think to look in this direction. Thanks!

Reply:
It is not fixing the GC crash. It is fixing a different crash that was being hit some of the time.